controllers/managedclusterForCGU_controller.go

```diff
@@ -0,0 +1,207 @@
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	clusterv1 "github.com/open-cluster-management/api/cluster/v1"
+	policiesv1 "github.com/open-cluster-management/governance-policy-propagator/api/v1"
+	ranv1alpha1 "github.com/openshift-kni/cluster-group-upgrades-operator/api/v1alpha1"
+)
+
+const (
+	clusterNotReadyRetryDelay = time.Minute * 5
+	clusterNoStatusRetryDelay = time.Minute * 1
+	ztpInstallNS              = "ztp-install"
+	ztpDeployWaveAnnotation   = "ran.openshift.io/ztp-deploy-wave"
+)
+
+// ManagedClusterReconciler reconciles a ManagedCluster object to auto create the ClusterGroupUpgrade
+type ManagedClusterForCguReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+// Reconcile the managed cluster auto create ClusterGroupUpgrade
+// - When a new managed cluster is created, create ClusterGroupUpgrade CR for the cluster only when it's ready
+//   and its child policies are available
+//
+// Note: The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ManagedClusterForCguReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	reqLogger := r.Log.WithValues("Request.Name", req.Name)
+	reqLogger.Info("Reconciling managedCluster to create clusterGroupUpgrade")
+
+	managedCluster := &clusterv1.ManagedCluster{}
+	err := r.Get(ctx, types.NamespacedName{Name: req.Name}, managedCluster)
+	if errors.IsNotFound(err) {
+		// managed cluster could have been deleted
+		return ctrl.Result{}, nil
+	}
+	if err != nil {
+		// Error reading the object, requeue the request
+		return ctrl.Result{}, err
+	}
+
+	clusterGroupUpgrade := &ranv1alpha1.ClusterGroupUpgrade{}
+	err = r.Get(ctx, types.NamespacedName{Name: req.Name, Namespace: ztpInstallNS}, clusterGroupUpgrade)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			// Error reading clusterGroupUpgrade, requeue the request
+			return ctrl.Result{}, err
+		}
+	} else {
+		// clusterGroupUpgrade for this cluster already exists, stop reconcile
+		r.Log.Info("clusterGroupUpgrade found", "Name", clusterGroupUpgrade.Name, "Namespace", clusterGroupUpgrade.Namespace)
+		return ctrl.Result{}, nil
+	}
+
+	availableCondition := meta.FindStatusCondition(managedCluster.Status.Conditions, clusterv1.ManagedClusterConditionAvailable)
+	if availableCondition == nil {
+		r.Log.Info("cluster has no status yet", "Name", managedCluster.Name, "RequeueAfter:", clusterNoStatusRetryDelay)
+		return ctrl.Result{RequeueAfter: clusterNoStatusRetryDelay}, nil
+	} else if availableCondition.Status == metav1.ConditionTrue {
+		// cluster is ready
+		policies, err := r.getPolicies(ctx, managedCluster.Name)
```

Comment:
We should add a comment clarifying when policies become available. Is there a race condition here (at what point does ACM copy the policies into the child namespace)? If ACM waits until the cluster is ready to start creating child policies we could end up with a partial copy. If they are populated when the ManagedCluster is created (with the right labels) we should be able to avoid the race condition as install takes 45+ minutes.

Reply:
The child policies actually get created as soon as the policy/placementrule/placementbinding are created. 

Thanks Ian for verifying/testing this on a hub cluster.  I tried it by myself as well by creating fake cluster(It will never be ready), and then creating policy/placementrule/placmentbinding. The child policy is created immediately when policy/placementrule/placmentbinding are created.

As we query the child policies after cluster is ready, this gives ACM ~45mins to create child policies, so we shouldn't have race condition here. I will update the code to remove the additional waiting.

Reply:
This means that so long as the user creates policies at approximately the same time as they create the site there should be no race condition here. :)